src/rules/spaceBeforeFunctionParenRule.ts

```diff
@@ -0,0 +1,173 @@
+    this._visitFunction(node);
+    super.visitFunctionDeclaration(node);
+  }
+
+  protected visitFunctionExpression(node: ts.FunctionExpression): void {
+    this._visitFunction(node);
+    super.visitFunctionExpression(node);
+  }
+
+  public visitMethodDeclaration(node: ts.MethodDeclaration): void {
+    this._visitFunction(node);
+    super.visitMethodDeclaration(node);
+  }
+
+  public visitMethodSignature(node: ts.SignatureDeclaration): void {
+    this._visitFunction(node);
+    super.visitMethodSignature(node);
+  }
+
+  private _visitFunction(node: ts.Node): void {
+    const options = this.getOptions();
+
+    if (options === null || options.length === 0 || Object.keys(options).length === 0) {
+      return;
+    }
+
+    let identifier: ts.Node;
+    let openParen: ts.Node;
+    let isAsyncArrow = false;
+
+    const isArrow = node.kind === ts.SyntaxKind.ArrowFunction;
+
+    node.getChildren().forEach((child: ts.Node): void => {
+      switch (child.kind) {
+        case ts.SyntaxKind.SyntaxList:
+          if (isArrow && child.getStart() === node.getStart() && child.getText() === "async") {
+            isAsyncArrow = true;
+          }
+          break;
+
+        case ts.SyntaxKind.Identifier:
+        case ts.SyntaxKind.ConstructorKeyword:
+          identifier = child;
+          break;
+
+        case ts.SyntaxKind.OpenParenToken:
+          openParen = child;
+          break;
+
+        default: break;
+      }
+    });
+
+    if (!openParen || (isArrow && !isAsyncArrow)) {
+      return;
+    }
+
+    const hasIdentifier = identifier && (identifier.getEnd() !== identifier.getStart());
+    let optionName;
+
+    if (isAsyncArrow) {
+      optionName = "asyncArrow";
+    } else {
+      switch (node.kind) {
+        case ts.SyntaxKind.Constructor:
+          optionName = "constructor";
+          break;
+
+        case ts.SyntaxKind.MethodDeclaration:
+        case ts.SyntaxKind.MethodSignature:
+          optionName = "method";
+          break;
+
+        default:
+          optionName = hasIdentifier ? "named" : "anonymous";
+      }
+    }
+
+    const option = this.getOption(optionName);
+    this.scanner.setTextPos(openParen.getStart() - 1);
+    const prevTokenKind = this.scanner.scan();
+    const hasSpace = prevTokenKind === ts.SyntaxKind.WhitespaceTrivia;
+    let failure: Lint.RuleFailure;
+
+    if (hasSpace && option === "never") {
+      const fix = new Lint.Fix(Rule.metadata.ruleName, [
+        this.deleteText(openParen.getStart() - 1, 1),
+      ]);
+      failure = this.createFailure(openParen.getStart(), 1, Rule.INVALID_WHITESPACE_ERROR, fix);
+    } else if (!hasSpace && option === "always") {
+      const fix = new Lint.Fix(Rule.metadata.ruleName, [
+        this.appendText(openParen.getStart(), " "),
+      ]);
+      failure = this.createFailure(openParen.getStart(), 1, Rule.MISSING_WHITESPACE_ERROR, fix);
+    }
+
+    if (failure) {
```

Comment:
please avoid implicit type coercion to boolean -- do an explicit `undefined` check instead